### ISSUE TITLE
Consider a `Client::ReadRow()` function.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp-spanner#386]
Reading a single row is a very common operation. Right now it requires a lot of boiler plate code from the application, it would be convenient to refactor this to a function we provide.

**Comment by coryan** [comment]
I think we can push this to the Beta release because it is just adding functionality.

**Comment by devjgm** [comment]
Q: How would a caller know that the results will be exactly one row and thus that this function would be safe to use?

**Comment by coryan** [comment]
Typically the query would be `"SELECT * FROM table WHERE primary_key = @value"`, that is, the application is restricting by a primary key, so there is only one value returned.  I guess things like `SELECT COUNT(*) FROM table` are also possible candidates for this function.

**Comment by devjgm** [comment]
OK, those examples make sense. I'm trying to think about this feature would interact w/ existing functions/classes that are already part of the API. So maybe rather than adding this as a method on `Client`, maybe we should add it as a convenience function on `ResultSet`:

```cc
class ResultSet {
 public:
  ...
  template <typename RowType>
  RowParser<RowType> Rows();  // EXISTING FUNCTON

  template <typename RowType>
  StatusOr<RowType> ConsumeSingleRow();  // PROPOSED (modulo naming)
};
``

So, a caller might use this in as follows:

```cc
  Client c = ...
  auto results = c.ExecuteSql("SELECT ... one single row");
  if (!result) { ... }
  auto row = results.ConsumeSingleRow<RowType>();
  if (row) {
    auto [x, y, z] = row->get();
    ...
  }
```

Question: Would this makes the existing code meaningfully simpler? I.e., is this a worthwhile simplification for a common-enough use case???

**Comment by devjgm** [comment]
There is a refactoring coming up to support `SELECT *` (#387) and the API changes resulting from that may affect what we do here. This issue is basically blocked on that issue.

**Comment by devjgm** [comment]
FWIW, we now have a nice place to put a method like this if we decide we want one: [RowStream](https://github.com/googleapis/google-cloud-cpp-spanner/blob/7412cde8ed6268ae8f17d3dfb84588e0caf95d86/google/cloud/spanner/results.h#L65-L92)

On that class, we could potentially add a method like this:

```cc
class RowStream {
 public:
  ...
  StatusOr<Row> NextRow();
  ...
};
```
(or maybe we should add it as a non-member function, since it's at best a convenience thing???)

And callers could use it (if they know that they'll *only* be getting a single row) like this:

```cc
// PROPOSED
RowStream rows = client.Read(...);
if (auto row = rows.NextRow()) {
  // Use *row
}
```

This is as opposed to what callers might have to write today, using a loop, which would be:

```cc
// TODAY
RowStream rows = client.Read(...);
for (auto row : rows) {
  if (row) {
    // Use *row
    break;
  }
}
```

The proposed code is a bit shorter and easier to read for the case where the user wants to read and process a single row.

Questions:

1. Do we think this is a common enough operation that it needs a convenience function?
2. Should we add such a convenience function now or wait until we have more experience with the library (and maybe some user feedback?

Thoughts?

**Comment by devjgm** [comment]
I lean toward not adding a `ReadOneRow()` functionality at this point for the following reasons:

1. It's at best a convenience. It's not adding any new functionality that doesn't exist today; just a convenience.

2. I'm not sure reading a single row is all that common. A quick skim through all of our samples.cc, and I didn't see any that clearly return a single row (but I admit, I may have missed cases). My point is, I'm not sure it's all that *common*.

3. Programmers *knowing* that a query is guaranteed to return a single row often requires external reasoning about the schema of the table. That is, it requires that the caller may know that due to the schema and the keys used in the query, only a single row can possibly be returned. But even that reasoning may change in the future if the schema changes. It seems safer to encourage users to use a loop to consume all the returned rows.

4. This is an easy feature to add in a future non-breaking change. So I don't see any need to add it now for beta.

So I'm leaning toward closing this, and maybe re-opening it later if we have new data that suggests this feature would be more useful than it seems like now.

Other thoughts? Counter points?

**Comment by coryan** [comment]
> I lean toward not adding a `ReadOneRow()` functionality at this point for the following reasons:
> 
> 1. It's at best a convenience. It's not adding any new functionality that doesn't exist today; just a convenience.

That is true for the whole library.

> 4. This is an easy feature to add in a future non-breaking change. So I don't see any need to add it now for beta.

That I think the winning argument.

> Other thoughts? Counter points?

Let's remove this from the milestone, but leave it open in case folks want to vote for it.

**Comment by devjgm** [comment]
SGTM. Dropping from Beta. Leaving open so others can see, cite, vote, etc.

**Comment by devjgm** [comment]
OK. So I think @coryan was right. After writing several more samples, there are cases where it would be nice to have a "get me one row" function. I have two examples so far. So I think I'm going to propose a PR.
